pkg/api/pss_test.go

```diff
@@ -0,0 +1,393 @@
+func TestPssSend(t *testing.T) {
+	var (
+		logger = logging.New(ioutil.Discard, 0)
+
+		mtx             sync.Mutex
+		recievedTargets trojan.Targets
+		recievedTopic   trojan.Topic
+		recievedBytes   []byte
+		done            bool
+
+		sendFn = func(_ context.Context, targets trojan.Targets, topic trojan.Topic, bytes []byte) error {
+			mtx.Lock()
+			recievedTargets = targets
+			recievedTopic = topic
+			recievedBytes = bytes
+			done = true
+			mtx.Unlock()
+			return nil
+		}
+
+		pss          = newMockPss(sendFn)
+		client, _, _ = newTestServer(t, testServerOptions{
+			Pss:    pss,
+			Storer: mock.NewStorer(),
+			Logger: logger,
+		})
+
+		targets   = fmt.Sprintf("[[%d]]", 0x12)
+		topic     = "testtopic"
+		hasher    = swarm.NewHasher()
+		_, err    = hasher.Write([]byte(topic))
+		topicHash = hasher.Sum(nil)
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("err - bad targets", func(t *testing.T) {
+		jsonhttptest.Request(t, client, http.MethodPost, "/pss/send/to/badtarget", http.StatusBadRequest,
+			jsonhttptest.WithRequestBody(bytes.NewReader(payload)),
+			jsonhttptest.WithExpectedJSONResponse(jsonhttp.StatusResponse{
+				Message: "Bad Request",
+				Code:    http.StatusBadRequest,
+			}),
+		)
+	})
+
+	t.Run("ok", func(t *testing.T) {
+		jsonhttptest.Request(t, client, http.MethodPost, "/pss/send/testtopic/12", http.StatusOK,
+			jsonhttptest.WithRequestBody(bytes.NewReader(payload)),
+			jsonhttptest.WithExpectedJSONResponse(jsonhttp.StatusResponse{
+				Message: "OK",
+				Code:    http.StatusOK,
+			}),
+		)
+		waitDone(t, &mtx, &done)
+		if !bytes.Equal(recievedBytes, payload) {
+			t.Fatalf("payload mismatch. want %v got %v", payload, recievedBytes)
+		}
+		if targets != fmt.Sprint(recievedTargets) {
+			t.Fatalf("targets mismatch. want %v got %v", targets, recievedTargets)
+		}
+		if string(topicHash) != string(recievedTopic[:]) {
+			t.Fatalf("topic mismatch. want %v got %v", topic, string(recievedTopic[:]))
+		}
+	})
+}
+
+// TestPssPingPong tests that the websocket api adheres to the websocket standard
+// and sends ping-pong messages to keep the connection alive.
+// The test opens a websocket, keeps it alive for 500ms, then receives a pss message.
+func TestPssPingPong(t *testing.T) {
+	var (
+		pss, cl, _ = newPssTest(t, opts{pingPeriod: 90 * time.Millisecond})
+
+		msgContent = make([]byte, len(payload))
+		tc         swarm.Chunk
+		mtx        sync.Mutex
+		pongWait   = 1 * time.Millisecond
+		done       = make(chan struct{})
+	)
+
+	cl.SetReadLimit(swarm.ChunkSize)
+	err := cl.SetReadDeadline(time.Now().Add(pongWait))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer close(done)
+	go waitReadMessage(t, &mtx, cl, msgContent, done)
+
+	m, err := trojan.NewMessage(topic, payload)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tc, err = m.Wrap(targets)
+	if err != nil {
+		t.Fatal(err)
+	}
+	<-time.After(500 * time.Millisecond) // wait to see that the websocket is kept alive
```

Comment:
This is the same as calling time.Sleep.

Reply:
personal preference to using channels :) anyway i changed to Sleeps